### PR TITLE
TTM get model changes

### PR DESCRIPTION
**Pull request by ajati**
The PR contains:

Changes in the `get_model` function to support any context length and forecast length.

### CODE REVIEW

File: tsfm_public/toolkit/gluonts_data_wrapper.py [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+            if np.isnan(out[i, ...]).any():
+                raise Exception("NaN found in forecast!")
+        return out
+
+
+class TorchDatasetFromGluonTSTrainingDataset(Dataset):
+    def __init__(
+        self,
+        gluon_dataset: TrainingDataset,
+        seq_len: int,
```
Comment (wgifford) [inline tsfm_public/toolkit/gluonts_data_wrapper.py:111]: Are these choices consistent with GluonTS? Do we want to introduce different conventions in our repository?

File: tsfm_public/toolkit/ddp_utils.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
+import os
+from datetime import timedelta
+
+import torch
+
+
+def init_ddp(timeout=600):
```
Comment (wgifford) [inline tsfm_public/toolkit/ddp_utils.py:7]: Is this needed?

File: tsfm_public/toolkit/gluonts_data_wrapper.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+            for i in range(target.shape[0]):
+                target[i, ...] = LastValueImputation()(target[i, ...])
+        elif len(target.shape) == 1:
+            target = LastValueImputation()(target)
+        else:
+            raise Exception("Only 1D and 2D arrays are accepted by the impute_series() function.")
+    return target
+
+
+def np_to_torch(np):
```
Comment (wgifford) [inline tsfm_public/toolkit/gluonts_data_wrapper.py:26]: Similar function exists in `datasets.py`

File: tsfm_public/toolkit/gluonts_data_wrapper.py [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+    if np.dtype == "float" or np.dtype == "float32":
+        return torch.from_numpy(np).float()
+    elif np.dtype == "int":
+        return torch.from_numpy(np)
+
+
+def _torch(*nps):
```
Comment (wgifford) [inline tsfm_public/toolkit/gluonts_data_wrapper.py:33]: Similar function may exist in `dataset.py`

File: tsfm_public/toolkit/gluonts_data_wrapper.py [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+
+        Returns:
+            Iternator: With each series transformed.
+        """
+        assert len(self.mean) > 0
+        assert len(self.std) > 0
+
+        out = list(data)
+        for i, _ in tqdm(enumerate(out)):
+            out[i]["target"] = (impute_series(out[i]["target"]) - self.mean[i]) / (self.std[i])
```
Comment (wgifford) [inline tsfm_public/toolkit/gluonts_data_wrapper.py:84]: Do we want to check for small standard deviation?

File: tsfm_public/toolkit/gluonts_data_wrapper.py [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+        return torch.from_numpy(np).float()
+    elif np.dtype == "int":
+        return torch.from_numpy(np)
+
+
+def _torch(*nps):
+    return tuple(np_to_torch(x) for x in nps)
+
+
+class StandardScalingGluonTSDataset:
```
Comment (wgifford) [inline tsfm_public/toolkit/gluonts_data_wrapper.py:37]: Is this multi-series aware?

File: tsfm_public/toolkit/gluonts_data_wrapper.py (hunk 1 above)
Comment (wgifford) [reply tsfm_public/toolkit/gluonts_data_wrapper.py:111]: Is it possible to leverage our existing torch datasets?

File: tsfm_public/toolkit/visualization.py [hunk 7]
```diff
@@ -309,7 +309,7 @@ def plot_predictions(
                 if k in signature_keys:
                     random_samples[k] = torch.stack([dset[i][k] for i in indices]).to(device=device)
             output = model(**random_samples)
-            predictions_subset = output.prediction_outputs[:, :, channel].squeeze().cpu().numpy()
```
Comment (wgifford) [inline tsfm_public/toolkit/visualization.py:312]: Are we sure this doesn't break other uses?

**Review by wgifford — CHANGES_REQUESTED** [review]

File: tsfm_public/toolkit/gluonts_data_wrapper.py (hunk 3 above)
Comment (ajati) [reply tsfm_public/toolkit/gluonts_data_wrapper.py]: Thanks. Using those. https://github.com/ibm-granite/granite-tsfm/pull/219/commits/3598090ba2dca854d9af8226e1b9e814b3471026

File: tsfm_public/toolkit/gluonts_data_wrapper.py (hunk 4 above)
Comment (ajati) [reply tsfm_public/toolkit/gluonts_data_wrapper.py]: Thanks. https://github.com/ibm-granite/granite-tsfm/pull/219/commits/3598090ba2dca854d9af8226e1b9e814b3471026

**Comment by ajati** [comment]
@wgifford @vijaye12 Please review the updated PR.

File: tsfm_public/toolkit/get_model.py [hunk 8]
```diff
@@ -93,28 +94,51 @@ def get_model(
...
+                if force_return:
+                    selected_context_length = available_context_lens[-1]
+                    LOGGER.warning(
+                        f"Requested context length is too short. Requested = {context_length}. Available lengths for model_type = {model_path_type} are: {available_context_lens}. Returning the shortest context length model possible since `force_return=True`. Data needs to be handled properly, and it can affect the performance!"
+                    )
+                else:
+                    raise ValueError(
+                        f"Requested context length is too short. Requested = {context_length}. Available lengths for model_type = {model_path_type} are: {available_context_lens}. To return the shortest context length model possible, set `force_return=True`."
+                    )
+
```
Comment (wgifford) [inline tsfm_public/toolkit/get_model.py:141]: Minor nit in the exception below. Can we use `ValueError("In the current implementation, set `freq_prefix_tuning` to None for automatic model selection accordingly.")`

**Review by vijaye12 — APPROVED** [review]

File: tsfm_public/toolkit/get_model.py (hunk 8 above)
Comment (ajati) [reply tsfm_public/toolkit/get_model.py:141]: Done. Thanks.

**Pull request merged**